summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -1,102 +1,238 @@
+    provider: ComponentProvider,
+    batcher: Batcher,
+    shards: KeyValueShards
+  ): Topology = {
+    val component = provider.createSingle[(K, V), Sharded[K, V]](wrapSharded(batcher, shards))
+    val (componentId, topologyWithComponent) = topology.withComponent(getNodeName(node), component)
+    registerShardedEdges[K, V](topologyWithComponent, node, componentId)
   }
 
-  private def getSummerId(node: SummerNode[Storm]): Topology.BoltId[_, _] =
+  private def registerItemEdges[T](
+    topology: Topology,
+    source: StormNode,
+    sourceId: Topology.EmittingId[Item[T]]
+  ): Topology = stormDag.dependantsOf(source).foldLeft(topology) {
+    case (currentTopology, downstreamNode: FlatMapNode[Storm]) =>
+      currentTopology.withEdge(Edges.shuffleItemToItem[T](
+        sourceId,
+        getFMBoltId[T](downstreamNode),
+        withLocalGrouping(downstreamNode)
+      ))
+    case (_, downstreamNode: SummerNode[Storm]) =>
+      throw new Exception(s"Impossible to create item edge to summer node: " +
+        s"$sourceId -> ${getNodeName(downstreamNode)}")
+  }
+
+  private def registerAggregatedEdges[K, V](
+    topology: Topology,
+    source: StormNode,
+    sourceId: Topology.EmittingId[Aggregated[K, V]]
+  ): Topology = stormDag.dependantsOf(source).foldLeft(topology) {
+    case (_, downstreamNode: FlatMapNode[Storm]) =>
+      throw new Exception(s"Impossible to create aggregated edge to flat map node: " +
+        s"$sourceId -> ${ getNodeName(downstreamNode) }")
+    case (currentTopology, downstreamNode: SummerNode[Storm]) =>
+      currentTopology.withEdge(Edges.groupedAggregatedToSummer[K, V](
+        sourceId,
+        getSummerBoltId[K, V](downstreamNode)
+      ))
+  }
+
+  private def registerShardedEdges[K, V](
+    topology: Topology,
+    source: StormNode,
+    sourceId: Topology.EmittingId[Sharded[K, V]]
+  ): Topology = stormDag.dependantsOf(source).foldLeft(topology) {
+    case (currentTopology, downstreamNode: FlatMapNode[Storm]) =>
+      currentTopology.withEdge(Edges.shuffleShardedToItem[K, V](
+        sourceId,
+        getFMBoltId[(K, V)](downstreamNode),
+        withLocalGrouping(downstreamNode)
+      ))
+    case (currentTopology, downstreamNode: SummerNode[Storm]) =>
+      currentTopology.withEdge(Edges.groupedShardedToSummer[K, V](
+        sourceId,
+        getSummerBoltId[K, V](downstreamNode)
+      ))
+  }
+
+  private def getSummerBoltId[K, V](node: SummerNode[Storm]): SummerBoltId[K, V] =
     Topology.BoltId(getNodeName(node))
 
-  private def getFlatMapId(node: FlatMapNode[Storm]): Topology.BoltId[_, _] =
+  private def getFMBoltId[T](node: FlatMapNode[Storm]): FlatMapBoltId[T] =
     Topology.BoltId(getNodeName(node))
 
-  private def getSourceId(node: SourceNode[Storm]): Topology.SpoutId[_] =
-    Topology.SpoutId(getNodeName(node))
+  private def shouldEmitKeyValues(node: StormNode): Boolean =
+    stormDag.dependantsOf(node).exists(_.isInstanceOf[SummerNode[_]])
```

Comment:
Yes, in future the other case for `groupedJoin` will be added here, for this review - only for summers.